File: south-africa/2016/scrunched.tle.cpp
/**
 * This program uses the triangle method to compute the arclength.
 * It finds the amplitude of the sine wave with binary search over real numbers.
 */
#include <iostream>
#include <iomanip>
#include <cmath>

using namespace std;

const auto pi = 3.141592653589793238462643383279502884L;
long double a, v, w;
int n;

long double f(long double x)
{
    return a * sin(2 * pi * n * x / v);
}

long double computeArclength()
{
    auto x = 0.0L;
    auto l = 0.0L;
    auto dx = 0.0001L;

    while (x <= v / (4 * n)) {
        auto dy = fabsl(f(x) - f(x + dx));
        l += sqrt(dx * dx + dy * dy);
        x += dx;
    }

    return l;
}

int main()
{
    long double l;

    while (cin >> w >> v >> n) {
        auto lo = 0.0L;
        auto hi = sqrt(w * w - v * v) / (2 * n);
        auto target = w / n;
        auto eps = 0.1L;

        while (eps > 0.0001L) {
            a = (lo + hi) / 2;
            l = computeArclength() * 4;
            if (l >= target) {
                hi = a;
            } else {
                lo = a;
            }
            eps = fabs(target - l);
        }
        cout << setprecision(3) << fixed << a - fmod(a, 0.001) << endl;
    }

    return 0;
}

File: south-africa/2016/scrunched.tle.py
from math import fabs, fmod, sqrt, sin, pi

def f(x):
    global a, n, v
    return a * sin(2 * pi * n * x / v)

def compute_arclength():
    x = 0
    l = 0.0
    dx = 0.0001
    while x <= v / (4 * n):
        dy = fabs(f(x) - f(x + dx))
        l += sqrt(dx * dx + dy * dy)
        x += dx
    return l


tokens = input().split(maxsplit=2)
while len(tokens) == 3:
    w, v, n = float(tokens[0]), float(tokens[1]), int(tokens[2])
    lo = 0
    hi = sqrt(w ** 2 - v ** 2) / (2 * n)
    target = w / n
    eps = 0.1
    while eps > 0.00001:
        a = (lo + hi) / 2
        l = compute_arclength() * 4
        if l >= target:
            hi = a
        else:
            lo = a
        eps = fabs(target - l)
    print("{:.3f}".format(a - fmod(a, 0.001)))
    tokens = input().split(maxsplit=2)
